interval.hh

```diff
@@ -88,22 +89,156 @@ public:
+    bool _end_exists = false;
+    bool _end_inclusive = false;
+    bool _singular = false;
+    union {
+        T _start_value;
+    };
+    union {
+        T _end_value;
+    };
 public:
-    wrapping_interval(optional<bound> start, optional<bound> end, bool singular = false)
-        : _start(std::move(start))
-        , _singular(singular) {
+    wrapping_interval(optional<bound> start, optional<bound> end, bool singular = false) {
+        _singular = singular;
+        if (start) {
+            _start_exists = true;
+            _start_inclusive = start->is_inclusive();
+            std::construct_at(&_start_value, std::move(start->value()));
+        }
         if (!_singular) {
-            _end = std::move(end);
+            if (end) {
+                _end_exists = true;
+                _end_inclusive = end->is_inclusive();
+                try {
+                    std::construct_at(&_end_value, std::move(end->value()));
+                } catch (...) {
+                    if (_start_exists) {
+                        std::destroy_at(&_start_value);
+                    }
+                    throw;
+                }
+            }
+        }
+    }
+    wrapping_interval(T value) {
+        _start_exists = true;
+        _start_inclusive = true;
+        _singular = true;
+        std::construct_at(&_start_value, std::move(value));
+    }
+    wrapping_interval(const wrapping_interval& o)
+            : _start_exists(o._start_exists)
+            , _start_inclusive(o._start_inclusive)
+            , _end_exists(o._end_exists)
+            , _end_inclusive(o._end_inclusive)
+            , _singular(o._singular) {
+        if (_start_exists) {
+            std::construct_at(&_start_value, o._start_value);
+        }
+        if (_end_exists) {
+            try {
+                std::construct_at(&_end_value, o._end_value);
+            } catch (...) {
+                if (_start_exists) {
+                    std::destroy_at(&_start_value);
+                }
+                throw;
+            }
+        }
+    }
+    wrapping_interval(wrapping_interval&& o) noexcept(std::is_nothrow_move_constructible_v<T>)
+            : _start_exists(o._start_exists)
+            , _start_inclusive(o._start_inclusive)
+            , _end_exists(o._end_exists)
+            , _end_inclusive(o._end_inclusive)
+            , _singular(o._singular) {
+        if (_start_exists) {
+            std::construct_at(&_start_value, std::move(o._start_value));
+        }
+        if (_end_exists) {
+            try {
+                std::construct_at(&_end_value, std::move(o._end_value));
+            } catch (...) {
+                if (_start_exists) {
+                    std::destroy_at(&_start_value);
+                }
+                throw;
+            }
+        }
+    }
+    wrapping_interval& operator=(const wrapping_interval& o) {
+        if (this != &o) {
+            if (_start_exists) {
+                std::destroy_at(&_start_value);
```

Comment:
Shouldn't we do the same dance for exception safety like in the move-assignment? To avoid double-destroy, and destroy-without-construct in case of exception later in this function?

Reply:
You mean clear _start_exists? yes. I'll add tests.

Reply:
No, we can't throw here. But we do need to improve the code later down.